builtin/providers/cloudstack/resource_cloudstack_ssh_keypair.go

```diff
@@ -0,0 +1,120 @@
+package cloudstack
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+func resourceCloudStackSSHKeyPair() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudStackSSHKeyPairCreate,
+		Read:   resourceCloudStackSSHKeyPairRead,
+		Delete: resourceCloudStackSSHKeyPairDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"public_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"private_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"fingerprint": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceCloudStackSSHKeyPairCreate(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	name := d.Get("name").(string)
+	public_key, ok := d.GetOk("public_key")
+
+	if ok && public_key != "" {
+		//register key supplied
+		p := cs.SSH.NewRegisterSSHKeyPairParams(name, public_key.(string))
+		r, err := cs.SSH.RegisterSSHKeyPair(p)
+		if err != nil {
+			return err
+		}
+		log.Printf("[DEBUG] RegisterSSHKeyPair response: %+v\n", r)
+		log.Printf("[DEBUG] Key pair successfully registered at Cloudstack")
+		d.SetId(name)
+	} else {
+		//no key supplied, must create one and return the private key
+		p := cs.SSH.NewCreateSSHKeyPairParams(name)
+		r, err := cs.SSH.CreateSSHKeyPair(p)
+		if err != nil {
+			return err
+		}
+		log.Printf("[DEBUG] CreateSSHKeyPair response: %+v\n", r)
+		log.Printf("[DEBUG] Key pair successfully generated at Cloudstack")
+		log.Printf("[DEBUG] Private key returned: %s", r.Privatekey)
+		d.Set("private_key", r.Privatekey)
+		d.SetId(name)
+	}
+
+	return resourceCloudStackSSHKeyPairRead(d, meta)
+}
+
+func resourceCloudStackSSHKeyPairRead(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	log.Printf("[DEBUG] looking for ssh key  %s with name %s", d.Id(), d.Get("name").(string))
+	p := cs.SSH.NewListSSHKeyPairsParams()
+	p.SetName(d.Get("name").(string))
+	r, err := cs.SSH.ListSSHKeyPairs(p)
+	if err != nil {
+		return err
+	}
+	if r.Count == 0 {
+		log.Printf("[DEBUG] Key pair %s does not exist", d.Get("name").(string))
+		d.Set("name", "")
+		return nil
+	} else {
```

Comment:
Since the previous `if` body contains a return at the end, there is no need to write the `else` here. In Go it's more common to emit these if possible: https://golang.org/doc/effective_go.html#if